### ISSUE TITLE
Bug in indexing + reductions

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire#1005]
```
#include <arrayfire.h>
#include <cstdio>
#include <cstdlib>

using namespace af;

int main(int argc, char *argv[])
{
    try {

        array a = constant(0, 75, 75, 5, f32);
        for (int i = 0; i < 5; i++) {
            array tmp = randu(75, 75, f32);
            a(span, span, i) = tmp;

            float val0, val1;
            unsigned idx0, idx1;

            min<float>(&val0, &idx0, a(span, span, i));
            min<float>(&val1, &idx1, tmp);

            printf("%f, %f\n", val0, val1);
            printf("%d, %d\n", idx0, idx1); // should be same but are not
        }


    } catch (af::exception& e) {
        fprintf(stderr, "%s\n", e.what());
        throw;
    }

    return 0;
}
```

**Comment by pavanky** [comment]
Might be related to #981

**Comment by pavanky** [comment]
Only seems to be a problem with indices. Values seem to be correct.

**Comment by pavanky** [comment]
This is caused by two issues:

- On OpenCL the look up names for different template parameters were causing a conflict.

- On CUDA + OpenCL, the index values are coming out incorrectly for

**Comment by shehzan10** [comment]
Fixed by PR #1009
